### ISSUE TITLE
Publisher/caller disclosure

### ISSUE THREAD

**Issue by oberstet** [wamp-proto/wamp-proto#57]
Provide options so we can provide any of:
- WAMP session ID
- **authrole** of WAMP session
- **authid** of WAMP session

**Comment by mbonneau** [comment]
:+1:

I would like to see the ```disclose_caller```/```disclose_publisher``` option provide all this information if the router has it.

**Comment by darkl** [comment]
:+1: 
Can this be added to the spec? It seems that [Thruway already supports this](https://github.com/voryx/Thruway/blob/b6d62af2aa2acd4389463766804c40d5ef6c3a86/src/Thruway/Call.php#L303), and I think it could be every useful for callees.

**Comment by oberstet** [comment]
@darkl Crossbar.io had implemented that in the past. And we removed it again. The reason: the next thing after authid/authrole asked for was transport level details: "we need access to the WebSocket cookie in callees". CB has more information about the Caller than just session ID, authid and authrole. Where do we stop? It doesn't make sense to transmit all that information in each and every call.

So regarding the question "What info should be disclosed", here are the options:

1. only `sessionid`
2. `sessionid`, `authid` and `authrole`
3. `sessionid`, `authid`, `authrole`, `transport_details`, ...

Now it was suggested in the past to let the Subscriber/Callee not only ask for "details yes/no", but ask for specific details ("I need only authrole. Gimme that and nothing else").

The problem with that is: when different Subscribers ask for different details, the router cannot get away with serializing an event **once** - as the details asked for differ. The same problem arises with Callees and shared registrations.

**Comment by goeddea** [comment]
Since we want to keep the routing core performant, we probably don't want to think about having configurable details.

Absent this, providing `sessionid`, `authid` and `authrole` seems to cover a large number of use cases (this has been requested a few times) while not increasing wire traffic by too much. 

My suggestion is to go with this slightly expanded set and close the issue.

**Comment by oddjobz** [comment]
Just a thought, how about something like "disclose_my_transport" as an additional option in the client? Or, just going out on a limb, "disclose_fields=[..]" ?? if that's too fine-grained, then maybe four options, "nothing", "session", "session,auth,role", "kitchen sink" ... ??

**Comment by oberstet** [comment]
@oddjobz no, because, see above.

**Comment by oddjobz** [comment]
@oberstet , Ok, in principle I can see that if the client is allowed to choose the fields it wants returned, then when you publish an event, you may need to send different identification fields to each client, and I appreciate there is an undesirable performance hit if you need to manipulate the packet to each client separately depending on their requirements.

However .. in practice (Ok, so I'm speaking about "my" code so this may be a flawed perspective) the code that calls a given topic on the client side, will be common to all clients, so the client identification requirements will 'always' be the same for a given topic. Whereas I may well re-use the same topic in different places in the code, again the identification requirements will always remain the same as this is a function of the topic implementation.

i.e. If the topic "needs" the "auth" in order to function, then the client must always disclose the "auth", or the topic call will fail.

I guess I'm looking at this "the other way around", the required fields are determined by the topic implementation and not by the client, yet currently the fields available are specified by the client. Perhaps (?) we need a handshake here, the client specifies the details it's prepared to disclose, and the server specifies the details it "needs" to function (?)

On the first invocation the client may send more than is needed, which can be filtered out by Crossbar, then on return the server can notify the client of the subset of fields it needs, so subsequent calls can be 100% efficient .. for example many of my RPC implementations need "auth" only and not "session" or anything else .. so this scheme would produce a notable reduction in traffic over sending session/auth/role for everything. 

This adds to the complexity, but on the other hand it does give you an additional layer of error checking in terms of handshaking required data .vs. available data, and assuming you need to pass more than just the session by default, this will minimise the traffic flow.

If you take into account the additional complexity being added (well, that I find I need) to cope with the fact I'm "just" getting a session, the performance / complexity difference may be self-cancelling, and indeed it shifts complexity away from user code and back onto infrastructure .. which IMHO would be the correct direction of travel.

Alternatively (?) if "disclose_me" was relative to the Websocket connection, then each topic implementation specified the fields required ... don't know .. but the current implementation seems to be application centric to a degree in that for what I'm doing, I don't actually want a client-side disclose_me "at all" as security is handled server side (it just gets in the way). 

Given all the client-side code is sent "from" the server, "disclose_me" has no security function, and in terms of bandwidth the choice is dependent on the topic's requirements ... sorry, I'm going in circles, does that make sense?

**Comment by goeddea** [comment]
@oddjobz - Just a couple of points: 

* The client currently authorizes the disclosure of information, but it does not send this. Any information about the caller is added by the router. 
* The current spec and implementation of having the caller decide whether its identity should be disclosed has rightfully been criticized (see e.g. https://github.com/crossbario/crossbar/issues/283). We'll likely not keep it that way.

**Comment by oddjobz** [comment]
@goeddea 

>The client currently authorizes the disclosure of information ..
Sure, appreciate that.

>The current spec and implementation ..
Ok, from my perspective, bandwidth is an issue because the link between the router and the topic server is always a network connection and will have to carry all the variables required, at the moment this is just "session", but the point of "where to draw the line" is a very good one.

Whereas I appreciate there are different use-cases at play here, for me the client-side disclose_me has zero value with regards to security. Yet the inability to get immediate access to the variables I *need* is very expensive in terms of development resource and indeed back-end processing in terms of making those variables available. Currently I need to maintain a session listing in a shared database so topic implementations can resolve sessions back to authid's. I know there are metadata calls to do this but I'm currently not happy with is this mechanism.

For what it's worth I'm still thinking an "I need .." definition in the topic implementation might be a good solution ...

**Comment by goeddea** [comment]
@oddjobz - I think drawing the line at sessionID, authID and authrole makes sense, since this is what most people need. 

Regarding the configuration of where to trigger the disclose: For RPCs this could equally be the callee when it registers the procedure or a configuration in the router. For PubSub with multiple subscribers we run into the serialization issue again - one subscriber requests the details,  the others don't want them, two serializations. In this case having it in the router configuration and set for the entire application might make more sense.

**Comment by oddjobz** [comment]
Sure, but again I guess it's use-case dependent. For me, all subscribers would have the same requirements, so having it set by the publisher for all subscribers would be fine. I could imagine that the same publication may attract two subscribers who want different details .. but in practice, how often would this happen?

**Comment by nbrochu** [comment]
If I'm mostly interested in the Callee's IP address, is this the best option with the current state of the protocol?

```python
# Anywhere inside an ApplicationSession
self._transport.http_headers.get("x-real-ip") or \
self._transport.http_headers.get("x-forwarded-for") or \
self._transport.peer.split(":")[1]
```

I was looking to update http://stackoverflow.com/questions/26725770/accessing-rpc-callers-ip-and-http-connect-headers-inside-applicationsessions-r but looking at this issue, it looks like passing the Caller Transport was only in for a single release.

**Comment by meejah** [comment]
@nbrochu something like:

```
    @inlineCallbacks
    def onJoin(self, _):

        @inlineCallbacks
        def method(deets):
            session = yield self.call('wamp.session.get', details.caller)
            print "session_info:", session
            peer = session['transport']['peer']
            print "peer's address", peer
            headers = session['transport']['http_headers_received']
            print "headers:"
            print '\n'.join(['{}: {}'.format(k, v) for (k, v) in headers.items()])

        yield self.register(
            method, 'some_method',
            types.RegisterOptions(details_arg='details''),
        )
```

I haven't tried this exact code behind nginx etc to confirm the x-real-ip etc headers.

**Comment by oddjobz** [comment]
@meejah just as a matter of interest, what features of NGINX are you using? i.e. what is driving the use of crossbar 'behind' NGINX?

**Comment by meejah** [comment]
@oddjobz I'm not, but @nbrochu's question implied he was (``X-Forwarded-For`` etc).

**Comment by nbrochu** [comment]
I wouldn't worry about it. That was my setup a long time ago, before Crossbar, when I had a standalone Autobahn Component. I would proxy pass the web socket connection in NGINX for TLS.

That being said, when Crossbar ends up getting multi-node support, some load-balancing setups could  use some form of reverse-proxying and this becomes relevant again.

I also know that some people like doing more advanced forms of A/B testing with NGINX: Proxy pass to stable or experimental based on a cookie value (generally). With Crossbar though, it makes more sense to simply connect to a different realm with some frontend logic in the browser.

**Comment by oddjobz** [comment]
Mmm, have you tried "ipvs" ?

**Comment by oberstet** [comment]
This is how it is implemented in Crossbar.io now:

1. Both publisher and caller disclosure can be enabled on a per URI basis in the router configuration
2. The detail attributes in EVENT messages are: `publisher` (the WAMP session ID of the publisher), `publisher_authid` and `publisher_authrole`
3. The detail attributes in the INVOCATION messages are: `caller` (the WAMP session ID of the caller), `caller_authid` and `caller_authrole`

I do think this reflects the end result of our discussion of the feature. @mbonneau does it?

**It isn't reflected in the specification though.** And eg Thruway uses different attribute names (see [here](https://github.com/crossbario/autobahn-js/issues/219).

**Comment by oddjobz** [comment]
Hi, is there any documentation on this yet? (do you have a link?)

**Comment by oberstet** [comment]
@oddjobz rgd Crossbar.io, no, the docs are not there=( https://github.com/crossbario/crossbar/issues/841

**Comment by oberstet** [comment]
Essentially, the semantics have converged now, the feature is well defined - but we need to go over the spec text.

**Comment by gammazero** [comment]
Only referenced discussion, no re-opening issue.

**Comment by gammazero** [comment]
I understand that the consensus here is to include the `authid` and `authrole` fields in support of publisher and caller identification.  To clarify, what is the agreed name for these fields:
`publisher_authid`, `publisher_authrole`, `caller_authid`, `caller_authrole` or simply `authid` and `authrole`.

To me, it seems that the `publisher_` or `caller_` prefix is unnecessary, as this is apparent from the message type in which they are delivered.  I think this will only require a minor exit to the existing spec text.

Also, is there any agreed formatting for the custom `x_...` fields.  Looks like the convention is to have `x_` then some abbreviation of the implementation, such as `tw_` or `cb_` and then the field name.  Any agreement here?

**Comment by darkl** [comment]
@gammazero, unfortunately the agreed names include the prefix, i.e. `publisher_authid`, `publisher_authrole`, `caller_authid`, `caller_authrole`. Please don't change this again, as this is already reflected in the implementations (see [here](https://github.com/Code-Sharp/WampSharp/blob/wampv2/src/net45/WampSharp/WAMP2/V2/Core/Contracts/Rpc/InvocationDetails.cs#L58), [here](https://github.com/Code-Sharp/WampSharp/blob/wampv2/src/net45/WampSharp/WAMP2/V2/Core/Contracts/PubSub/EventDetails.cs#L44) and [here](https://github.com/crossbario/autobahn-python/blob/d0e5cb841bb0a5cc48353b36a8578753100e8db1/autobahn/wamp/protocol.py#L934)).

**Comment by gammazero** [comment]
@darkl Thank you for the clarifying - will not change (the close-reopen was accidental).  I was not sure about decision as it looked like Thruway did not include the role prefix: https://github.com/voryx/ThruwayCommon/blob/master/src/Message/EventMessage.php#L149-L150

[nexus](https://github.com/gammazero/nexus) router now also supports these fields for caller and publisher identification as per your feedback.

**Comment by oberstet** [comment]
FWIW, there is a deeper running feature request somewhere .. I think filed by @mbonneau : allow to assigned one or multiple authroles when authenticating a client .. which is sth we should discuss IMO ..

**Comment by mbonneau** [comment]
@gammazero Thruway is a little out of sync with the spec.

@oberstet Would still like to see the `authroles` officially supported see #99

**Comment by oberstet** [comment]
@mbonneau I am +1 taking a vote on #99 : I think this has clear use cases, but also has deep implementation impacts

**Comment by om26er** [comment]
Just a reminder, this is still out of sync with the spec. Crossbar now uses `details=true` in register options, however Nexus still uses `disclose_caller`. The spec text still uses `disclose_caller`. https://wamp-proto.org/wamp_latest_ietf.html#section-11.6

@KSDaemon what are your thoughts on this ?

**Comment by oberstet** [comment]
> Crossbar now uses `details=true` in register options

This is a client side feature: the decision to provide caller details in an invocation of to the callee, or to provide callee details to the caller in a result is purely router side

You can see this here

https://github.com/crossbario/autobahn-python/blob/528f6a05c2091e67a7b1cf7d968901119ec924df/autobahn/wamp/types.py#L901

which extract the attributes from RegisterDetails which _are_ forwarded in a REGISTER message. `details` and `details_arg` are not forwarded.

---

however, you are indeed right: 

[11.6. ](https://wamp-proto.org/wamp_latest_ietf.html#section-11.6)[Caller Identification](https://wamp-proto.org/wamp_latest_ietf.html#name-caller-identification)

it seem it is still reflecting the old approach (where caller/callee could request this ..) ...

**Comment by oberstet** [comment]
> * Caller Identification feature seems to be supported in many implementations

yes. the question though is: how is it controlled?

historically, caller identification was controlled by the caller, hence the "disclose_me" name.

the other option is: the router (configuration) always controls caller identification. and only the router.

fwiw, crossbar only supports the latter .. for at least 6 years .. https://github.com/crossbario/crossbar-examples/blob/3dbda8d44f4c8ac261714a97fd94f6d6462743c1/disclose/.crossbar/config.json#L22

**Comment by ecorm** [comment]
> Personally I prefer `disclose_me` naming rather than `details` because the latter is a rather general term: what details? `disclose_me` is also maybe not so perfect, but it refers to info about calling peer.

Yeah, I was also scratching my head when I saw `details=true`. I had no clue what that meant. It's especially confusing seeing that `details` is a field in several WAMP message types.

> historically, caller identification was controlled by the caller, hence the "disclose_me" name.
> 
> the other option is: the router (configuration) always controls caller identification. and only the router.

Those options need not be mutually exclusive. Caller identification could be controlled by the caller by default, or the router can override that decision.

**Comment by oberstet** [comment]
> Those options need not be mutually exclusive. Caller identification could be controlled by the caller by default, or the router can override that decision.

actually, yeah, this is true. fwiw, I like that=)

e.g. crossbar could allow to configure caller/publisher disclosure: "true", "false", "choose"

and with "true"/"false", the router decides, while with "choose", the caller/publisher MAY choose to disclose itself

in node config terms, with a static realms config, in crossbar it could look like:

```
"realms": [
            {
               "name": "realm1",
               "roles": [
                  {
                     "name": "anonymous",
                     "permissions": [
                        {
                           "uri": "",
                           "match": "prefix",
                           "allow": {
                              "call": true,
                              "register": true,
                              "publish": true,
                              "subscribe": true
                           },
                           "disclose": {
                              "caller": "choose",
                              "publisher": "choose"
                           },
                           "cache": true
                        }
                     ]
                  }
               ]
            }
         ],
```

**Comment by Jopie64** [comment]
Why would it need to be the caller or router who decides to disclose the caller? It's the callee who wants to receive this info or not, and actually can do something with it. It's no use for the caller or router to disclose caller info when the callee is going to ignore it anyway...

I'd suggest to put the disclose caller option in the REGISTER command. Maybe allow the caller or router to block it for privacy reasons or something. But passing it anyway even when not asked for sounds weird to me...

**Comment by ecorm** [comment]
> I'd suggest to put the disclose caller option in the REGISTER command.

I like this idea. What do you think of a disclose publisher option in the SUBSCRIBE command?

> Maybe allow the caller or router to block it for privacy reasons or something.

In environments where the router is not trusted, the caller/publisher could only _ask_ that disclosure be blocked, and it would be up to the router to honor that request.

**Comment by oberstet** [comment]
My view:

> Why would it need to be the caller or router who decides to disclose the caller?

The router should be the only party that decides, because of security reasons. it would amount to the peers deciding about their own authorization, which would be a security design bug

> I'd suggest to put the disclose caller option in the REGISTER command.

No, it is the router that should decide, neither the callee, nor the caller.

**Comment by ecorm** [comment]
> It would amount to the peers deciding about their own authorization, which would be a security design bug

Deciding our own security is exactly what our backend application (a WAMP peer) was doing with Crossbar. Our backend app, not Crossbar, owns the user database, along with its roles and permissions schemes.

> No, it is the router that should decide, neither the callee, nor the caller.

We ended up configuring Crossbar to disclose everywhere, and our backend app simply ignores the session id, authid, and authrole details when they are not needed, depending on the RPC. It would otherwise have been a maintenance headache to configure this on a per-RPC basis in Crossbar.

**Comment by oberstet** [comment]
> Deciding our own security is exactly what our backend application (a WAMP peer) was doing with Crossbar.

You can use Dynamic Authorization, which is simply a WAMP component called in the right places by Crossbar.io

https://github.com/crossbario/crossbar-examples/tree/master/authorization/dynamic/embedded

Besides dynamically returning authorization based on WAMP URI and action (on a realm-role), you can enable caching (also selectively and dynamically), and you can go even further

https://github.com/crossbario/crossbar-examples/tree/master/payload-validation

This expands the info a dynamic authorizer may return by payload type information. Crossbar.io can automatically validate application payloads of calls or call returns against a schema definition. This works with any supported serializer.

**Comment by oberstet** [comment]
probably I should just add an example, maybe easier to grasp. you can configure a realm so that for authorizing WAMP actions in Crossbar.io, the router will call out to _your_ WAMP procedure

```python
class ExampleAuthorizer(ApplicationSession):

    @wamp.register('com.example.authorize')
    def authorize(self, session, uri, action, options):
        if session['authrole'] == 'frontend' and action == 'call' and \
           uri == 'com.example.get_candle_history':

            authorization = {
                'allow': True,
                'disclose': False,
                'validate': {
                    'call': 'trading.Period',
                    'call_result': 'trading.CandleResult',
                    'call_error': 'trading.ErrorInvalidPeriod'
                },
                'cache': True,
            }
            return authorization

        else:
            return False
```

The dict returned can contain the same knobs as you would use in a static realm configuration. Above example also uses `validate`, which is for payload validation. Since caching is also enabled, this will lead to:

* when a WAMP URI-action for a realm-role is to be performed the _first_ time, the router will call into your procedure
* it will then cache the authorization info, including `disclose`, and including the payload validation target types
* those validation types are then dynamically resolved (and again cached), and the app payload in the WAMP action is then also matched against the type. If it doesn't conform to the payload type desired, the router will deny the WAMP action

**Comment by ecorm** [comment]
> You can use Dynamic Authorization, which is simply a WAMP component called in the right places by Crossbar.io

Yes, I'm aware of this, but there was something preventing us from using it. I think it was the cache not being useful because user permissions can change at runtime in our app.

**Comment by oberstet** [comment]
> I think it was the cache not being useful because user permissions can change at runtime in our app.

Not sure if we have added it already for the authorization cache in particular, but eg for the cookie store (all stores work the same), we now have a LMDB persistent version which can be shared and used from all workers and nodes that can access the LMDB mmap'ed file.

The stores also have a run-time WAMP API themself, that is you can invalidate the persistently cached records from you app ...

It all comes together lately, rgd features and catching up in crossbar actually providing them. The dynamic authenticators and authorizers, similar to the meta API, and other stuff is all "eating our own dog food". They are all WAMP based and allow users to plug in their app code into the control plane of crossbar. Anyways, just saying;) If your application works as is, using your approach, great!

**Comment by oberstet** [comment]
> I think it was the cache not being useful because user permissions can change at runtime in our app.

yes, server-side logout is one thing, client-side active logout is the other. in this case the authentication store in crossbar will automatically delete the cached authentication record.

eg leaving a session in https://github.com/crossbario/crossbar-examples/blob/master/authentication/cookie/client.py with

```
self.leave('wamp.close.logout', message='some custom message that nobody uses and could be removed')
```

will delete the authentication for the stored cookie (it will not delete the data, just the auth .. by design)

**Comment by Jopie64** [comment]
> The router should be the only party that decides, because of security reasons. it would amount to the peers deciding about their own authorization, which would be a security design bug

Indeed authorization is an important aspect I failed to mention.
But there I think more of it this way:
It's the callee who determines whether it needs caller info, communicated by the REGISTER command. And it's the router who decides whether it honors that request or not, based on permissions.

An open question for me would be, when the router cannot honor the caller disclosure request, would it just leave caller info out the invocation, or would it deny the REGISTER command altogether?

I currently feel this is also up to the callee. Some callees _require_ caller info to function as in @ecorm s example. Others might just use it for statistics or something. Maybe a `required: true` option?

> What do you think of a disclose publisher option in the SUBSCRIBE command?

I'd consider that similar as with REGISTER.

But, as I read through the thread, I saw an argument against making it optional at the callee/subscriber side: in that case it could be that the router could not serialize a broadcast message only once since it needs to make a special message for every subscriber. This causes a potential performance penalty.

I think however that this performance argument doesn't weigh enough against the benefits. I don't think this will happen very much in practice especially in the REGISTER case since most of the time an invocation has to be send to only one node. Also I think there are multiple ways to fix it. One of which is to serialize only once for the node requesting the most disclosures and use the same message for other nodes who will probably ignore the extra info anyway.

What do you think?

**Comment by ecorm** [comment]
> Yes, I'm aware of this, but there was something preventing us from using it. I think it was the cache not being useful because user permissions can change at runtime in our app.

There was another motivation: to avoid added complexity. We would've had to maintain our own centralized registry of RPCs for use with the dynamic authorizer. The way we have it now is that the RPC handlers know what permissions are required and simply reject the request if the disclosed user does not have permission. No dynamic authorization needed and no worries about a cache becoming stale when permissions change at runtime.

**Comment by oberstet** [comment]
> It's the callee who determines whether it needs caller info, communicated by the REGISTER command. And it's the router who decides whether it honors that request or not, based on permissions.

actually, I think you've hit the sweet spot by pointing out: there are 2 parts logically, the peer _wants_ to get disclosure infos (and hence is ready to consume it, and the info isn't skipped in the client, and hence transferred without use), _and_ the router _authorizes_ (decides weather to allow that client to be served with included disclosure info or not)

for a specific client to receive disclosure infos, the client must then have requested that, _and_ the router must have granted receiving the infos

now, I agree, this is a nice and clean logical design.

if we want to go there, IMO, this needs a more careful analysis nevertheless. is it worth eg?

because, for example, a consequence of above would be: a router, when dispatching an EVENT to subscribers, will need to serialize 2 versions of the event (with and without disclosure), and then dispatch the right version to the respective subset of subscribers.

> I think however that this performance argument doesn't weigh enough against the benefits. 

for crossbar, above change would be maximal invasive, dig deep into the dispatching etc, and I can't see practical benefits in my use of crossbar. to the contrary. lots of risk.

the thing is:
* if a client doesn't understand disclosure infos, fine, simply ignore and not use the info in the client.
* if a client doesn't need disclosure infos, and the router nevertheless is configured to disclose (on some URI-action), then there are bytes wasted on those transports

for me, both of those consequences are "ok".

but yeah, agreed, on a logical level the **want+allow** design would be "cleaner" ... I don't want to disagree just because "it's work + risk" and "it's ok as is" ..

**Comment by ecorm** [comment]
> But, as I read through the thread, I saw an argument against making it optional at the callee/subscriber side: in that case it could be that the router could not serialize a broadcast message only once since it needs to make a special message for every subscriber. This causes a potential performance penalty.

Instead of disclosure being an option in a REGISTER or SUBSCRIBE command, it could be set on a per-procedure or per-topic basis via the router's meta API. That way disclosure is the same for all calls and publications. It also avoids the headache of maintaining a long list of RPCs//topics in a router's config file.

**Comment by oberstet** [comment]
> it could be set on a per-procedure or per-topic basis via the router's meta API.

crossbar allows you to do that already, but not via the WAMP meta API (which is _not_ concerned about router _control_), but via the crossbar control API (which is also WAMP). and currently only "per role" (hence you need to replace the role incl. permissions). so you 2 options:

* provide your own dynamic authorizer wamp component
* dynamically configure static roles (with static permissions)

so "static" is relative with the latter;) anyways, I started to describe a bit of that API here https://github.com/crossbario/autobahn-python/blob/528f6a05c2091e67a7b1cf7d968901119ec924df/autobahn/xbr/test/catalog/src/wamp-control.fbs#L27

and you can find the impl here https://github.com/crossbario/crossbar/blob/ccbae0a56ef331fe00192455837aa72578eadc54/crossbar/worker/router.py#L554  ...

**Comment by KSDaemon** [comment]
Hey, wanted to participate, but seems that you already got to the same conclusion :)

But anyway, my 50cents: I'm also vote for having this option on `Callee` side in `REGISTER`, as it is the only side that is interested in that data and knows how to proceed it. And yeah, having the possibility to control data disclosure on the router side — is good. 

Regarding events in topics: well, I think that we should not overcomplicate publications and serialize messages differently based on this option for every subscriber. The main idea of topics is that every subscriber receives exactly the same message (i would treat different event options as a useful part of the payload). So imho, sending disclosure details to every subscriber even if some of them are not interested in it — is ok.

**Comment by KSDaemon** [comment]
Regarding how to proceed when the callee expects and registers an RPC with disclose option set, but the router doesn't allow this... I think this should be handled on the callee side, as it is more business logic than protocol related.

**Comment by oberstet** [comment]
yeah, it seems like we can have it all .. more options;) fwiw, my current thinking, whatever we add in the spec, I think:

* the current option of the router (and only the router) deciding whether to disclose, and if so, include disclosure information for all callers/subscribers should be preserved (as default)
* other options, that is behavior where the disclosure is not determined only by the router, but may be influenced by the peers can be added, but we need to define the exact behavior (eg what if peer and router disagree etc)#

> more business logic than protocol related

in my eyes, disclosure is an attribute of authorization: given an URI and an action (e.g. publish), whether the client with authrole on the realm is allowed or not, and if so, whether disclosure information is to be added

authorization can also be dynamically determined by app code, yes, but that doesn't necessarily imply that the authorization would need to be done in one of the peers

my argument is: authentication and authorization of the app might or might not be done by app code itself, but _that_ AA app code should be confined and minimized for security reasons ..

**Comment by ecorm** [comment]
Submitted PR #502 so that the `publisher`, `publisher_authid`, `publisher_authrole`, `caller`, `caller_authid`, and `caller_authrole` properties are explicitly mentioned.
